aws/sign.go

```diff
@@ -309,16 +311,10 @@ func canonicalHeaders(sortedHeaderNames []string, host string, hdr http.Header)
 // lowercase hexadecimal string.
 func payloadHash(req *http.Request, hasher hasher) (string, error) {
```

Comment:
On second thoughts, I think a much simpler and less error prone approach would be to check in `payloadHash` if `req.Body` is an io.ReadSeeker. If it is, pass to hasher as-is and then seek back to the original pos, if it's not, use the old approach of reading into memory and then replacing req.Body.